### PR TITLE
fix Issue 17819 - static foreach segfaults on __traits(allMembers)

### PR DESCRIPTION
**Pull request by tgehr**
https://issues.dlang.org/show_bug.cgi?id=17819

This makes `static if` and `static foreach` handle "circular" dependencies the same way as string `mixin`s do it. (I think DMD uses a somewhat arbitrary definition of what constitutes a "circular" dependency.)
It is quite obvious that neither the current `static if` behavior nor the current string `mixin` behavior was consciously designed, so I don't know if this is a good fix. It is however the fix that was asked for.

### CODE REVIEW

File: src/dmd/attrib.d [hunk 1]
```diff
@@ -929,6 +928,7 @@ extern (C++) final class StaticIfDeclaration : ConditionalDeclaration
 {
     ScopeDsymbol scopesym;
     bool addisdone;
+    bool onStack = false;
```
Comment (JinShil) [inline src/dmd/attrib.d]: It took me a little while to understand what "onStack" means, but I see now.

**Review by JinShil — APPROVED** [review]

File: src/dmd/attrib.d (hunk 1 above)
Comment (jacob-carlborg) [inline src/dmd/attrib.d]: All new public symbols should have documentation.

Comment (tgehr) [reply src/dmd/attrib.d]: Added a comment.

Comment (tgehr) [reply src/dmd/attrib.d]: It does not need to be public, but neither does addisdone. Made both private and added comments.

**Comment by wilzbach** [comment]
(rebased to stable as it's a critical bug and should be included in the next point release)

File: src/dmd/attrib.d [hunk 2]
```diff
@@ -47,7 +47,6 @@ extern (C++) abstract class AttribDeclaration : Dsymbol
     {
         if (errors)
             return null;
-
```
Comment (RazvanN7) [inline src/dmd/attrib.d]: This change is unrelated to this PR

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/attrib.d (hunk 2 above)
Comment (tgehr) [reply src/dmd/attrib.d]: Which change?

**Comment by wilzbach** [comment]
How about adding the test from 18718?
https://github.com/wilzbach/dmd/blob/0abcf79eda89cea2409b484b07a84d33affd0a94/test/fail_compilation/test18718.d

**Comment by tgehr** [comment]
@wilzbach: This is already in there.

**Review by wilzbach — APPROVED** [review]
Sorry @tgehr I guess I need more sleep :O

File: src/dmd/attrib.d [hunk 3]
```diff
@@ -928,7 +927,8 @@ extern (C++) class ConditionalDeclaration : AttribDeclaration
 extern (C++) final class StaticIfDeclaration : ConditionalDeclaration
 {
     ScopeDsymbol scopesym;
-    bool addisdone;
+    private bool addisdone = false; // stores whether members have been added to scope
+    private bool onStack = false;   // stores whether a call to `include` is currently active
```
Comment (WalterBright) [inline src/dmd/attrib.d]: s/stores whether/true if/

**Review by WalterBright — APPROVED** [review]

File: src/dmd/attrib.d (hunk 3 above)
Comment (tgehr) [reply src/dmd/attrib.d]: Fixed.
